src/store/index-level.ts

```diff
@@ -1,213 +1,434 @@
+  }
+
+  /**
+   * Wraps the given operation as an operation for the specified index partition.
+   */
+  private async createOperationForIndexPartition(tenant: string, indexName: string, operation: LevelWrapperBatchOperation<string>)
+    : Promise<LevelWrapperBatchOperation<string>> {
+    // we write the index entry into a sublevel-partition of tenantPartition.
+    // putting each index entry within a sublevel allows the levelDB system to calculate a gt minKey and lt maxKey for each of the properties
+    // this prevents them from clashing, especially when iterating in reverse without iterating through other properties.
+    const tenantPartition = await this.db.partition(tenant);
+    const indexPartitionName = IndexLevel.getIndexPartitionName(indexName);
+    const partitionOperation = tenantPartition.createPartitionOperation(indexPartitionName, operation);
+    return partitionOperation;
+  }
+
+  /**
+   * Wraps the given operation as an operation for the itemId to indexes lookup partition.
+   */
+  private async createOperationForIndexesLookupPartition(tenant: string, operation: LevelWrapperBatchOperation<string>)
+    : Promise<LevelWrapperBatchOperation<string>> {
+    const tenantPartition = await this.db.partition(tenant);
+    const partitionOperation = tenantPartition.createPartitionOperation(INDEX_SUBLEVEL_NAME, operation);
+    return partitionOperation;
+  }
+
+  private static getIndexPartitionName(indexName: string): string {
+    // we create index partition names in __${indexName}__ wrapping so they do not clash with other sublevels that are created for other purposes.
+    return `__${indexName}__`;
+  }
+
+  /**
+   * Gets the index partition of the given indexName.
+   */
+  private async getIndexPartition(tenant: string, indexName: string): Promise<LevelWrapper<string>> {
+    const indexPartitionName = IndexLevel.getIndexPartitionName(indexName);
+    return (await this.db.partition(tenant)).partition(indexPartitionName);
+  }
+
+  /**
+   * Gets the itemId to indexes lookup partition.
+   */
+  private async getIndexesLookupPartition(tenant: string): Promise<LevelWrapper<string>> {
+    return (await this.db.partition(tenant)).partition(INDEX_SUBLEVEL_NAME);
+  }
+
+  /**
+   * Queries the index for items that match the filters. If no filters are provided, all items are returned.
+   *
+   * @param filters Array of filters that are treated as an OR query.
+   * @param queryOptions query options for sort and pagination, requires at least `sortProperty`. The default sort direction is ascending.
+   * @param options IndexLevelOptions that include an AbortSignal.
+   * @returns {string[]} an array of itemIds that match the given filters.
+   */
+  async query(tenant: string, filters: Filter[], queryOptions: QueryOptions, options?: IndexLevelOptions): Promise<string[]> {
+
+    // check if we should query using in-memory paging or iterator paging
+    if (IndexLevel.shouldQueryWithInMemoryPaging(filters, queryOptions)) {
```

Comment:
Hey @LiranCohen. I sanity overrode the `shouldQueryWithInMemoryPaging()` to always return `true`, just to test that things work regardless of the query strategy used, and 32 tests blew up.

I debugged a little and found that it seems that at least some tests are not passing any filter which caused the failures.

I think should we ether fixed the tests to have at least one filter so that the tests are not dependent on query strategy to pass, or change the code to accept no filter. What do you think?

We should probably optionally add a regression test to override `shouldQueryWithInMemoryPaging()` to make sure the test works on both strategy.

Reply:
Created a PR to address this: https://github.com/TBD54566975/dwn-sdk-js/pull/632, let me know what you think!

Reply:
🔥